File: stonequest/src/com/barelyconscious/game/player/activeeffects/Poison.java
/* *****************************************************************************
 * Project:          Roguelike2.0
 * File name:        Poison.java
 * Author:           Matt Schwartz
 * Date created:     07.25.2012
 * Redistribution:   You are free to use, reuse, and edit any of the text in
                     this file.  You are not allowed to take credit for code
                     that was not written fully by yourself, or to remove 
                     credit from code that was not written fully by yourself.  
                     Please email schwamat@gmail.com for issues or concerns.
 * File description: 
 **************************************************************************** */

package com.barelyconscious.game.player.activeeffects;

public class Poison extends Debuff {
    private float tickDamage; 
    private int tick;
    private int damageFrequency;
    
    public Poison(String name, int dur, float damage, int freq) {
        super(name, dur);
        tickDamage = damage;
        damageFrequency = freq;
        tick = damageFrequency;
    } // constructor
    
    // Countdown until the next damage tick
    public boolean nextTick() {
        if (--tick == 0) {
            tick = damageFrequency;
            return true;
        } // if
        
        return false;
    } // nextTick
    
    // How much damage the poison deals each tick
    public float getTickDamage() {
        return tickDamage;
    } // getTickDamage
    
    // How often the poison deals damage (in seconds)
    public int getDamageFrequency() {
        return damageFrequency;
    } // getDamageFrequency

    @Override
    public String toString() {
        return super.getName() + " inflicts " + (int)tickDamage + " damage.";
    } // toString
} // Poison

File: StoneQuest/src/com/barelyconscious/game/player/activeeffects/Poison.kt
package com.barelyconscious.game.player.activeeffects

import com.barelyconscious.game.Common
import com.barelyconscious.game.Game
import com.barelyconscious.game.graphics.LineElement
import com.barelyconscious.game.graphics.UIElement
import com.barelyconscious.systems.MessageSystem
import com.barelyconscious.services.TextLogMessageData
import com.barelyconscious.services.TextLogWriterService

/**
 * Create a new Poison with the following parameters
 *
 * @param name the displayName visible to the player of the Poison
 * @param dur how long the Poison lasts before it is removed
 * @param tickDamage the amount of damage dealt to the Player every freq tick
 * @param damageFrequency how often the Poison deals damage to the Player
 */
class Poison(
    name: String,
    dur: Int,
    val tickDamage: Double,
    val damageFrequency: Int,
    private val messageSystem: MessageSystem
) : Debuff(name, dur, Debuff.TOXIN, UIElement.POISON_ICON) {

    private var tick: Int = damageFrequency

    /**
     * Counts down until the next time the poison is to deal damage to the player
     *
     * @return true if the poison is to deal damage to the player on this tick
     */
    fun nextTick(): Boolean = if (--tick <= 0) {
        tick = damageFrequency
        true
    } else {
        false
    }

    override fun tick() {
        if (nextTick()) {
            Game.player.changeHealthBy(-tickDamage)

            messageSystem.sendMessage(
                TextLogWriterService.LOG_WRITE_TEXT,
                TextLogMessageData(toString())
                    .with(LineElement(displayName, true, Common.FONT_POISON_LABEL_RGB)),
                this)
        }

        super.tick()
    }

    /**
     * The message written to the TextLog when the Poison deals damage to the Player; the Player class actually writes
     * the message to the TextLog
     *
     * @return
     */
    override fun toString(): String {
        return "$displayName inflicts ${tickDamage.toInt()} damage."
    }
}
